src/cpu/dyn_cache.h

```diff
@@ -19,13 +19,26 @@
  *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
  */
 
+#include <cerrno>
 #include <cassert>
 #include <new>
 
 #include "mem_unaligned.h"
 #include "paging.h"
 #include "types.h"
 
+#if defined(__APPLE__)
+#include <AvailabilityMacros.h>
+#if defined(MAC_OS_VERSION_11_0) && (MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_VERSION_11_0)
+	#define HAVE_APPLE_WRITEPROT	1
+#endif
+#endif
```

Comment:
`HAVE_<function_name>` are usually wrangled out up in Meson, so would be nice to try to get it there.

Thinking about the this logic:  `MAC_OS_VERSION_11_0 && MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_VERSION_11_0`, when MAC_OS_VERSION_11_0 becomes MAC_OS_VERSION_11_1, it will stop working.
Same with `builtin_available(macOS 11.0.`

In Meson, can we directly looking for `pthread_jit_write_protect_np` and `sys_icache_invalidate` functions in the `libkern/OSCacheControl.h` headers? Then define corresponding `HAVE_PTHREAD_WRITE_PROTECT_NP` and `HAVE_SYS_ICACHE_INVALIDATE`, which follows the existing `HAVE_<exact_function_name>` approach.

Then, our preference for using the most optimal calls and falling back to less optimal - say;  HAVE_PTHREAD_WRITE_PROTECT_NP, then HAVE_MMAP, then HAVE_MPROTECT, (and so on) would simply follow the order of `#if, #elif, #elif, ... `

Reply:
I would definitely prefer to have this handled by Meson, and had honestly mostly copypasta'd that code from another project. I can investigate this further.

Reply:
Ah - a main problem here is that this needs to be determined at runtime, not sure moving it all to Meson will work.

Reply:
`MAC_OS_VERSION_11_0` will still be valid, since they get added to, e.g.:

```
#define MAC_OS_X_VERSION_10_14      101400
#define MAC_OS_X_VERSION_10_14_1    101401
#define MAC_OS_X_VERSION_10_14_4    101404
#define MAC_OS_X_VERSION_10_15      101500
#define MAC_OS_VERSION_11_0         110000
#define MAC_OS_VERSION_11_1         110100
```
You're really just using to look for "at least this version or higher".

Reply:
If we can directly check for the functions in Meson, then we should eliminate all of Apple's API baggage (`__APPLE__`, `MAC_OS_X_VERSION_X_Y..`, and  `if (__builtin_available(macOS 11.0, *).. and future-proof the code.  I also think we won't need these in Meson either. 

Maybe in MacOS 12.4, they come up with a couple even-more-recommended functions.. then no problem: we'll look for those in Meson just the same, and add their HAVE_<new_func> to the front of the #if, #else chain. 

I like that Apple has tried to name them somewhat Posix'y too - (instead of, say `apple_m1_cache_..`) - so if the day comes and BSD or Linux implements equivalent-named functions, then the code side of the house might actually stay the same (but that's just wishful thinking :-).

Reply:
But the kicker is the functions need to be tested for at runtime. For example, the binary may get built with SDK 11.1 for x86_64, but then gets run by someone using 10.15, and trying to use that call will fail.

Reply:
> Ah - a main problem here is that this needs to be determined at runtime, not sure moving it all to Meson will work.

I guess at a minimum, we need to retain the runtime `__builtin_available(macOS 11.0` check?

Reply:
> > Ah - a main problem here is that this needs to be determined at runtime, not sure moving it all to Meson will work.
> 
> I guess at a minimum, we need to retain the runtime `__builtin_available(macOS 11.0` check?
> 
> (Where as the block of code I flagged in this comment is all compile time, so hopefully it can move to Meson)

Yes, I think we can at least hoist that part into Meson, and retain the __builtin_available!